src/cmd/sign.js

```diff
@@ -63,42 +62,25 @@ export default function sign(
       getIdFromFile(savedIdPath),
     ]);
 
-    const manifestId = getManifestId(manifestData);
-
-    if (id && !manifestId) {
-      throw new UsageError(
-        `Cannot set custom ID ${id} - The add-on ID must be specified in the manifest.json file.`,
-      );
-    }
-    if (idFromSourceDir && !manifestId) {
+    const id = getManifestId(manifestData);
+    if (idFromSourceDir && !id) {
       throw new UsageError(
         'Cannot use previously auto-generated extension ID ' +
-          `${idFromSourceDir} - This add-on ID must be specified in the manifest.json file.`,
-      );
-    }
-    if (id && manifestId) {
-      throw new UsageError(
-        `Cannot set custom ID ${id} because manifest.json ` +
-          `already defines ID ${manifestId}`,
+          `${idFromSourceDir} - This extension ID must be specified in the manifest.json file.`,
       );
     }
-    if (id) {
-      log.info(`Using custom ID declared as --id=${id}`);
-    }
 
-    if (manifestId) {
-      id = manifestId;
-    }
+    if (!id) {
+      // We only auto-generate add-on IDs for MV2 add-ons on AMO.
+      if (manifestData?.manifest_version !== 2) {
```

Comment:
```suggestion
      if (manifestData.manifest_version !== 2) {
```

`manifestData` cannot be void. The existing `getManifestId(manifestData)` call above unconditionally reads properties off `manifestData`.